cas/ctg_store.c

```diff
@@ -2067,6 +2280,289 @@ int ctgdump(struct m0_motr *motr_ctx, char *fidstr, char *dump_in_hex_str)
+	} else {
+		*out = M0_CRV_INIT_NONE;
+		M0_LEAVE("ctg_op=%p, no version", ctg_op);
+	}
+}
+
+/*
+ * Synchronously inserts/updates a key-value record considering the version
+ * and the tombstone that were specified in the operation ("new_version")
+ * and the version+tombstone that were stored in btree ("old_version").
+ */
+static int versioned_put_sync(struct m0_ctg_op *ctg_op)
+{
+	struct m0_buf             *key    = &ctg_op->co_key;
+	struct m0_be_btree        *btree  = &ctg_op->co_ctg->cc_tree;
+	struct m0_be_btree_anchor *anchor = &ctg_op->co_anchor;
+	struct m0_be_tx           *tx     = &ctg_op->co_fom->fo_tx.tx_betx;
+	struct m0_crv              new_version = M0_CRV_INIT_NONE;
+	struct m0_crv              old_version = M0_CRV_INIT_NONE;
+	int                        rc;
+
+	M0_PRE(ctg_op->co_is_versioned);
+	M0_ENTRY();
+
+	ctg_op_version_get(ctg_op, &new_version);
+	M0_ASSERT_INFO(!m0_crv_is_none(&new_version),
+		       "Versioned PUT or DEL without a valid version?");
+
+	/*
+	 * TODO: Performance.
+	 * This lookup call can be avoided if the version comparison
+	 * was shifted into btree. However, at this moment the cost of
+	 * such a lookup is much lower than the cost of re-working btree API.
+	 * See the be_btree_search() call inside ::btree_save. It is the
+	 * point where btree_save does its own "lookup". Versions could
+	 * be compared right in there. Alternatively, btree_save could
+	 * use btree_cursor as a "hint".
+	 */
+	rc = M0_BE_OP_SYNC_RET(op,
+			       m0_be_btree_lookup_inplace(btree,
+							  &op,
+							  key,
+							  anchor),
+			       bo_u.u_btree.t_rc) ?:
+		ctg_vbuf_unpack(&anchor->ba_value, &old_version);
+
+	/* The tree is long-locked anyway. */
+	m0_be_btree_release(NULL, anchor);
+
+	if (!M0_IN(rc, (0, -ENOENT)))
+		return M0_ERR(rc);
+
+	/*
+	 * Skip overwrite op if the existing record is "newer" (version-wise)
+	 * than the record to be inserted. Note, <= 0 means that we filter out
+	 * the operations with the exact same version and tombstone.
+	 */
+	if (!m0_crv_is_none(&old_version) &&
+	    m0_crv_cmp(&new_version, &old_version) <= 0)
+		return M0_RC(0);
+
+	M0_LOG(M0_DEBUG, "Overwriting " CRV_F " with " CRV_F ".",
+	       CRV_P(&old_version), CRV_P(&new_version));
+
+	anchor->ba_value.b_nob = ctg_vbuf_packed_size(&ctg_op->co_val);
+	rc = M0_BE_OP_SYNC_RET(op,
+			       m0_be_btree_save_inplace(btree, tx, &op,
+							key, anchor, true,
+							ctg_op_zones(ctg_op)),
+			       bo_u.u_btree.t_rc);
+
+	if (rc == 0) {
+		ctg_vbuf_pack(&ctg_op->co_anchor.ba_value,
+			      &ctg_op->co_val,
+			      &new_version);
+
+		m0_ctg_state_inc_update(tx,
+					key->b_nob -
+					sizeof(struct generic_key) +
+					ctg_op->co_val.b_nob);
+	}
+
+	return M0_RC(rc);
+}
+
+/*
+ * Gets an alive record (without tombstone set) in btree.
+ * Returns -ENOENT if tombstone is set.
+ * Always sets co_out_ver if the record physically exists in the tree.
+ */
+static int versioned_get_sync(struct m0_ctg_op *ctg_op)
```

Comment:
How to get the record regardless it is alive or not.
Do we have such use case?
For example, the client user provides a version, and if the version is matched, (regardless the tbs is set or not), we can return the record back to client?

Reply:
It looks like this is a duplicate. I replied to the same question in the previous comment: https://github.com/Seagate/cortx-motr/pull/1515#discussion_r818493475

Reply:
Yes. Duplicated due to network issue and re-submit.